syntax = "proto2";

package rv.protos;

import "rastervision/protos/model_config.proto";
import "rastervision/protos/scene.proto";

message MakeChipsConfig {
    message ObjectDetectionOptions {
        /*
            The ratio of negative chips (those containing no bounding boxes) to
            positive chips. This can be useful if the statistics of the
            background is different in positive chips. For example, in car
            detection, the positive chips will always contain roads, but no
            examples of rooftops since cars tend to not be near rooftops.
            This option is not used when window_method is `sliding`.
        */
        required float neg_ratio = 1;

        /*
            When a box is partially outside of a training chip, it is not
            clear if (a clipped version) of the box should be included in
            the chip. If the IOA (intersection over area) of the box with the
            chip is greater than ioa_thresh, it is included in the chip.
        */
        optional float ioa_thresh = 2 [default=0.8];

        /*
            Different models in the Object Detection API have different inputs.
            Some models allow variable size inputs so several methods
            of building training data are required

            Valid values are:
                - chip (default)
                - label
                    - each label's bounding box is the positive window
                - image
                    - each image is the positive window
                - sliding
                    - each image is from a sliding window with 50% overlap
        */
        optional string window_method = 3 [default="chip"];

        /*
            If method is "label", the positive window can be buffered.
            If value is >= 0. and < 1., the value is treated as a percentage
            If value is >= 1., the value is treated in number of pixels
        */
        optional float label_buffer = 4 [default=0.];
    }

    message ClassificationOptions {

    }

    message SegmentationOptions {
        optional float debug_chip_probability = 1 [default=0.25];
        optional float empty_survival_probability = 2 [default=0.25];
        optional int32 number_of_chips = 3 [default=100];
    }

    message Options {
        optional int32 chip_size = 1;
        optional bool debug = 2 [default=true];

        // The directory where all the output should be placed.
        optional string output_uri = 3;

        oneof ml_options_type {
            ObjectDetectionOptions object_detection_options = 4;
            ClassificationOptions classification_options = 5;
            SegmentationOptions segmentation_options = 6;
        }
    }

    repeated Scene train_scenes = 1;
    repeated Scene validation_scenes = 2;
    required ModelConfig model_config = 3;
    required Options options = 4;
}
